### ISSUE TITLE
Can’t call StreamSupport.stream with wildcard-extends since 2.5.2

### ISSUE THREAD

**Issue by astiob** [typetools/checker-framework#2151]
Since version 2.5.2, this code fails to compile:

```java
class Test {
    <T> void foo(Iterable<? extends T> iterable) {
        Spliterator<? extends T> spliterator = iterable.spliterator();
        Stream<? extends T> stream = StreamSupport.stream(spliterator, false);

        Stream<? extends T> stream2 = StreamSupport.stream(iterable.spliterator(), false);
    }
}
```

Here are the error messages with `-AprintAllQualifiers`:

```
error: [type.argument.type.incompatible] incompatible types in type argument.
        Stream<? extends T> stream = StreamSupport.stream(spliterator, false);
                                                         ^
  found   : ? extends T extends @UnknownKeyFor Object
  required: [extends @UnknownKeyFor Object super @UnknownKeyFor null]
error: [type.argument.type.incompatible] incompatible types in type argument.
        Stream<? extends T> stream2 = StreamSupport.stream(iterable.spliterator(), false);
                                                          ^
  found   : ? extends T extends @UnknownKeyFor Object
  required: [extends @UnknownKeyFor Object super @UnknownKeyFor null]
```

Looking at the list of bugs closed in 2.5.2 (by the way, it would be nice to only list bugs that were actually fixed and omit those closed as duplicates and especially those closed because of lack of reduced test case), it seems like this might be caused by #370 or #1922.

I don’t see anything wrong with my code. First off, am I perhaps missing something? Is the checker actually right?

Since `StreamSupport` is not in the annotated JDK, I’ve tried adding a stub file for it, but using just `<T>` in the declaration of this method changes nothing and I can’t think of anything else to try instead. The classes and methods `Iterator`, `Iterator::spliterator`, `Spliterator` and `BaseStream` in the annotated JDK don’t have any explicit bounds on their type arguments, so I don’t know where the `super null` comes from (nor do I know how to undo it, or even what it really means). I’ve tried to pass an explicit type argument in the `stream` method call, but Java won’t let me specify a wildcard type here (as `StreamSupport.<? extends T>stream()`—presumably because it can’t know that I mean this wildcard to be the same as the other wildcard).

Is there any way to work around this short of modifying my method signature to take `Iterable<T>`? (Thankfully, in my particular case that is acceptable, and I can confirm it compiles successfully.)

**Comment by bmhm** [comment]
@astiob would you mind sharing your stub?

**Comment by astiob** [comment]
I don’t have a stub at the moment. If I recall correctly, I tried creating one that just copied the official signature for the `stream` method in hopes that the Checker Framework would infer the bounds differently from the stub source than it does from the class binary, but it didn’t help. I ended up changing my own method signature to take `Iterable<T>` rather than `Iterable<? extends T>`.

**Comment by bmhm** [comment]
I'm implementing the CacheLoader interface, which has:

`Map<K, V> loadAll(Iterable<? extends K> keys) throws CacheLoaderException;`

My Key is somtimes String, sometimes a immutable generated class.

I get this error:

```
[ERROR]   found   : ? extends String
[ERROR]   required: [extends Object super null]
```

**Comment by wmdietl** [comment]
I think my earlier comment https://github.com/typetools/checker-framework/issues/2151#issuecomment-418521410 was misleading and we realized that in the next meeting.
It might be enough to add a `StreamSupport` class to the annotated JDK. Then when generating the annotated JDK the correct defaults for KeyFor should be applied.

@astiob Did you try that route?

@bmhm Could you file a separate issue with a minimized example?

**Comment by astiob** [comment]
I have not tried modifying the annotated JDK, only adding a stub file. I fear that it might be hard to rebuild it, but that’s because I don’t know anything about it. Is it actually hard?

What does the `super null` actually mean?

**Comment by wmdietl** [comment]
In #2200 I added a `StreamSupport` class for nullness, which picks up the special defaults we use in the java packages.
I tried with your example and no longer get the nullness errors.
@astiob  Can you try whether this fixed your original issue?

`super null` is the lower bound for a type variable or wildcard, null being the bottom of the type hierarchy.

**Comment by wmdietl** [comment]
Also, https://checkerframework.org/manual/#map-key-defaults-lowerbound discusses the issues with Key For lower bounds defaults.
Please let us know if there is something we should improve in that section.
Thanks for the bug report!

**Comment by astiob** [comment]
I can confirm that the newly released version 2.5.7 of the Checker Framework compiles my original real-world code successfully. Thanks!
